zebra/rt_netlink.c

```diff
@@ -831,56 +832,54 @@ static void _netlink_route_build_singlepath(const char *routedesc, int bytelen,
 	 * you fix this assumption
 	 */
 	label_buf[0] = '\0';
-	/* outgoing label - either as NEWDST (in the case of LSR) or as ENCAP
-	 * (in the case of LER)
-	 */
-	nh_label = nexthop->nh_label;
-	if (rtmsg->rtm_family == AF_MPLS) {
-		assert(nh_label);
-		assert(nh_label->num_labels == 1);
-	}
 
-	if (nh_label && nh_label->num_labels) {
-		int i, num_labels = 0;
-		u_int32_t bos;
+	assert(nexthop);
+	for (struct nexthop *nh = nexthop; nh; nh = nh->rparent) {
```

Comment:
Shouldn't this be for (ALL_NEXTHOPS...)?

Reply:
`ALL_NEXTHOPS` loops through all nexthops of a given route. What we need here is something different. When a recursive nexthop is resolved, the labels of both the recursive route and the resolved route should be stacked and installed in the FIB. If the recursive route (e.g. BGP) is labeled but the resolved route (e.g. OSPF) is not, then the recursive route can not be installed in the FIB. This is because MPLS labels have local significance and only the nexthop of the recursive route knows how to interpret its label. Recursive MPLS labels are a fundamental building block of L3VPNs for example.

Back to your question, at this point of the code what we need is to follow the parent nexthop until we find the "root" recursive route (Bingen implemented support for multiple recursion levels a while ago).

This PR however doesn't address the problem completely. When the label of the resolved nexthop changes (e.g. LDP going wild), the recursive route is not notified and thus not updated in the FIB. I'll mark this PR as preliminary until I find a good solution for this (extend ZEBRA_NEXTHOP_UPDATE or introduce ZEBRA_LABEL_UPDATE).